_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+![`Msf::ModuleManager.wrapper_module` `until` loop](/images/msf-module-manager-wrapper-module-until.png)
+
+First, why am I starting with `Object` as the parent?  Well, it's because constant resolution all start with `Object.get_const`, which I got from `ModuleConstMissing#const_missing` (https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L176)
+
+![`ModuleConstMissing#const_missing` using `Object` as constant resolution root](/images/module-const-missing-const-missing-object.png)
+
+`until` is to `while` as `unless` is to `if`, so this `until` loop is walking down the descendants until it gets to a leaf, which must be the wrapper_module that should be returned.
+
+There are some slight complications that were encountered that weren't in the original constraints: some modules have components of their name which are leading digits or non-alphanumeric characters, such as '-'. To ensure a unique mapping, I hex encoded those illegal characters.  Since hex encoding is reversible, I know it is a 1-to-1 mapping and will maintain the uniqueness of the original names.  Hex encoding could return a digit again, which isn't allowed in a Module name, so I have to prefix hex encoded characters with `X`
+
+![`Msf::ModuleManager.wrapper_module` hex-encoding](/images/msf-module-manager-wrapper-module-hex-encoding.png)
+
+Module reloading is handled by removing the leaf constant if it already exist when the loop gets to it (`module_names.empty?` is `true`).    `false` is passed to `const_defined?` to speed constant lookup and to prevent false positives from constants defined in ancestors.
+
+![`Msf::ModuleManager.wrapper_module` leaf-removal](/images/msf-module-manager-wrapper-module-leaf-removal.png)
+
+Walking down the tree is handled here
+
+![`Msf::ModuleManager.wrapper_module` walk tree](/images/msf-module-manager-wrapper-module-walk.png)
+
+The `if` handles if we created this part of the hierarchy before, which would be the case during reload, but also for common parts of the Metasploit Module names like `auxiliary/pro/social_engineering`, which is used by multiple Metasploit Modules.
+
+![`Msf::ModuleManager.wrapper_module` child defined](/images/msf-module-manager-wrapper-module-child-defined.png)
+
+Finally we get back to creating an anonymous module.  Prefixing a constant with `::` tells Ruby to look up a constant starting at the top-level (Object) instead of the current lexical scope, which in this case is `Msf`, which contains `Msf::Module`.
+
+![`Msf::ModuleManager.wrapper_module` `Module.new`](/images/msf-module-manager-wrapper-module-module-new.png)
+
+So, we didn't pass the `child_name` to `Module.new`, how does it know it's name?  Well it's turns out that the Ruby VM tells a Module it's name the first time it's associated with a constant, so setting the `child_name` constant on the `parent` namespace module will make the Module be named.
```

Comment:
So, **if** we didn't pass the `child_name` to `Module.new`, how does it know it's name?  Well **~~it's~~** **it** turns out that the Ruby VM tells a Module it's name the first time it's associated with a constant, so setting the `child_name` constant on the `parent` namespace module will make the Module be named.